backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/ShipmentScheduleQtyReservedStorage.java

```diff
@@ -0,0 +1,134 @@
+	private final IShipmentScheduleEffectiveBL shipmentScheduleEffectiveBL = Services.get(IShipmentScheduleEffectiveBL.class);
+	private final IWarehouseDAO warehouseDAO = Services.get(IWarehouseDAO.class);
+	@NonNull final ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails;
+
+	public ShipmentScheduleQtyReservedStorage(final @NonNull ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails)
+	{
+		this.stockDetails = stockDetails;
+	}
+
+	@Override
+	public List<ShipmentScheduleAvailableStockDetail> getStockDetailsMatching(final @NonNull I_M_ShipmentSchedule sched)
+	{
+		if (stockDetails.isEmpty())
+		{
+			return Collections.emptyList();
+		}
+		else
+		{
+			//
+			// Reserved qty shall always be in VHUs, no need for picking BOM.
+			final StockDataQuery mainProductQuery = toQuery(sched);
+			return new ArrayList<>(getStockDetailsMatching(mainProductQuery));
+		}
+	}
+
+	private ImmutableList<ShipmentScheduleAvailableStockDetail> getStockDetailsMatching(@NonNull final StockDataQuery query)
+	{
+		return stockDetails
+				.stream()
+				.filter(stockDetail -> matching(query, stockDetail))
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	private static boolean matching(final StockDataQuery query, final ShipmentScheduleAvailableStockDetail stockDetail)
+	{
+		//
+		// Product
+		if (!ProductId.equals(query.getProductId(), stockDetail.getProductId()))
+		{
+			return false;
+		}
+
+		//
+		// Warehouse
+		final Set<WarehouseId> queryWarehouseIds = query.getWarehouseIds();
+		if (!queryWarehouseIds.isEmpty() && !queryWarehouseIds.contains(stockDetail.getWarehouseId()))
+		{
+			return false;
+		}
+
+		//
+		// Attributes
+		final boolean queryMatchesAll = query.getStorageAttributesKey().isAll();
+		final boolean queryMatchesStockDetail = AttributesKey.equals(query.getStorageAttributesKey(), stockDetail.getStorageAttributesKey());
+		return queryMatchesAll || queryMatchesStockDetail;
+	}
+
+	@Override
+	public StockDataQuery toQuery(final @NonNull I_M_ShipmentSchedule sched)
+	{
+		final TableRecordReference scheduleReference = TableRecordReference.ofReferenced(sched);
+		final Util.ArrayKey materialQueryCacheKey = Util.ArrayKey.of(
+				scheduleReference.getTableName(),
+				scheduleReference.getRecord_ID(),
+				I_M_ShipmentSchedule.Table_Name,
+				sched.getM_ShipmentSchedule_ID());
+
+		return cachedMaterialQueries.getOrLoad(materialQueryCacheKey, k -> toQuery0(sched));
+	}
+
+	private StockDataQuery toQuery0(@NonNull final I_M_ShipmentSchedule sched)
+	{
+		final WarehouseId shipmentScheduleWarehouseId = shipmentScheduleEffectiveBL.getWarehouseId(sched);
+		final Set<WarehouseId> warehouseIds = warehouseDAO.getWarehouseIdsOfSamePickingGroup(shipmentScheduleWarehouseId);
+
+		final ProductId productId = ProductId.ofRepoId(sched.getM_Product_ID());
+
+		return StockDataQuery.builder()
+				.warehouseIds(warehouseIds)
+				.productId(productId)
+				.build();
```

Comment:
why are we skipping the attributes here?

Reply:
I have no clue how to create an AttributesKey from ASI, nor is it needed to do so.